File: traumaIA.js
// traumaIA.js — IA para módulo TRAUMA (imagenología) con puntos dolorosos
// ESM (Node >= 18)
// Usa OpenAI si hay API key; si falla, usa fallback heurístico.

const OPENAI_API = "https://api.openai.com/v1/chat/completions";
const MODEL = process.env.OPENAI_MODEL || "gpt-4o-mini";

/* ------------------------------------------------------------- */
/* Utils de normalización y fallback                             */
/* ------------------------------------------------------------- */

function normalizarExamen(examen = "", dolor = "", lado = "") {
  let x = String(examen || "").trim();
  if (!x) return "";
  x = x.toUpperCase();
  if (!x.endsWith(".")) x += ".";

  // Añadir lateralidad si aplica
  const L = (lado || "").toUpperCase();
  const lat = L ? ` ${L}` : "";
  if (/\b(CADERA|RODILLA|HOMBRO|TOBILLO|PIERNA|BRAZO|CODO|MUÑECA|MANO|PIE)\b/.test(x) &&
      lat && !/\b(IZQUIERDA|DERECHA)\b/.test(x)) {
    x = x.replace(/\.$/, "") + lat + ".";
  }

  // Estandariza ECO partes blandas con zona
  if (/ECOGRAF[ÍI]A.*PARTES\s+BLANDAS/.test(x) && !/\bDE\b/.test(x)) {
    const zona = (dolor || "").toUpperCase();
    if (zona && !/COLUMNA/.test(zona)) {
      x = `ECOGRAFÍA DE PARTES BLANDAS DE ${zona}${lat}.`.toUpperCase();
    }
  }
  return x;
}

function fallbackHeuristico(p = {}) {
  const d = String(p.dolor || "").toLowerCase();
  const L = (p.lado || "").toUpperCase();
  const lat = L ? ` ${L}` : "";
  const mayor60 = Number(p.edad) > 60;

  if (d.includes("rodilla")) {
    const diagnostico = `Gonalgia${lat ? ` ${L.toLowerCase()}` : ""}`;
    const examen = mayor60
      ? `RX DE RODILLA${lat} AP/LATERAL/AXIAL.`
      : `RESONANCIA MAGNÉTICA DE RODILLA${lat}.`;
    const justificacion =
      "Estudio dirigido a estructuras intraarticulares y periarticulares de la rodilla. La elección prioriza rendimiento diagnóstico según edad y sospecha; RX evalúa compromiso óseo/degenerativo y RM valora meniscos, cartílago y ligamentos. Los puntos dolorosos orientan la sospecha específica.";
    return { diagnostico, examen, justificacion };
  }

  const diagnostico = "Dolor osteoarticular localizado";
  const examen = "EVALUACIÓN IMAGENOLÓGICA SEGÚN CLÍNICA.";
  const justificacion =
    "Se inicia estudio de la región comprometida, priorizando seguridad y costo-efectividad. La selección del examen se ajusta a la sospecha clínica y hallazgos dirigidos.";
  return { diagnostico, examen, justificacion };
}

/* ------------------------------------------------------------- */
/* Mapeo clínico: puntos de rodilla → hipótesis                  */
/* ------------------------------------------------------------- */

/**
 * Espera { frente: string[], lateral: string[], posterior: string[] }
 * con etiquetas como: "interlínea medial/lateral", "rótula", "tuberosidad tibial",
 * "pes anserino", "banda iliotibial/gerdy", "fosa poplítea", etc.
 */
function inferirHipotesisDesdePuntos(rodillaMarcadores = {}) {
  const labels = [
    ...(rodillaMarcadores.frente || []),
    ...(rodillaMarcadores.posterior || []),
    ...(rodillaMarcadores.lateral || []),
  ]
    .map((s) => String(s || "").toLowerCase())
    .filter(Boolean);

  const hit = (rx) => labels.some((t) => rx.test(t));

  const hallazgos = [];
  const sospechas = [];

  // Interlíneas → meniscos
  if (hit(/\binterl[ií]nea?\s+medial\b/)) {
    hallazgos.push("Dolor selectivo en interlínea medial");
    sospechas.push("Lesión de menisco medial");
  }
  if (hit(/\binterl[ií]nea?\s+lateral\b/)) {
    hallazgos.push("Dolor selectivo en interlínea lateral");
    sospechas.push("Lesión de menisco lateral");
  }

  // Rótula / PF
  if (hit(/\b(r[óo]tula|patelar|patelofemoral|ap(e|é)x)\b/)) {
    hallazgos.push("Dolor peripatela/patelofemoral");
    sospechas.push("Síndrome patelofemoral o condropatía rotuliana");
  }

  // Tuberosidad tibial
  if (hit(/\btuberosidad\s+tibial\b/)) {
    hallazgos.push("Dolor en tuberosidad tibial");
    sospechas.push("Osgood–Schlatter o tendinopatía del tendón rotuliano");
  }

  // Pes anserino
  if (hit(/\b(pes\s+anserin[oó]|pata\s+de\s+ganso)\b/)) {
    hallazgos.push("Dolor en región pes anserino");
    sospechas.push("Tendinopatía del pes anserinus / bursitis anserina");
  }

  // Banda iliotibial / Gerdy
  if (hit(/\b(gerdy|banda\s+ilio?tibial|tracto\s+ilio?tibial)\b/)) {
    hallazgos.push("Dolor sobre inserción del tracto iliotibial");
    sospechas.push("Síndrome de la banda iliotibial");
  }

  // LCM / LCL (aproximación por bordes)
  if (hit(/\bborde\s+medial\b/) || hit(/\bepic[óo]ndilo\s+femoral\s+medial\b/)) {
    hallazgos.push("Dolor en trayecto medial de la rodilla");
    sospechas.push("Esguince del ligamento colateral medial (LCM)");
  }
  if (hit(/\bborde\s+lateral\b/) || hit(/\bepic[óo]ndilo\s+femoral\s+lateral\b/)) {
    hallazgos.push("Dolor en trayecto lateral de la rodilla");
    sospechas.push("Esguince del ligamento colateral lateral (LCL)");
  }

  // Fosa poplítea
  if (hit(/\bpopl[ií]tea?\b/)) {
    hallazgos.push("Dolor/firmeza en fosa poplítea");
    sospechas.push("Quiste de Baker / patología poplítea");
  }

  if (hallazgos.length === 0) hallazgos.push("Dolor selectivo en puntos marcados");
  if (sospechas.length === 0) sospechas.push("Compromiso meniscal o periarticular a precisar");

  return { hallazgos, sospechas };
}

function describirPuntosClinicos(detalles = {}) {
  const m = detalles?.rodillaMarcadores || null;
  if (!m) return { texto: "Sin puntos dolorosos marcados.", hallazgos: [], sospechas: [] };

  const { hallazgos, sospechas } = inferirHipotesisDesdePuntos(m);
  const texto = [
    "PUNTOS DOLOROSOS MARCADOS",
    ...hallazgos.map((h) => `• ${h}`),
    "",
    "SOSPECHAS ORIENTADAS POR LOS PUNTOS",
    ...sospechas.map((s) => `• ${s}`),
  ].join("\n");

  return { texto, hallazgos, sospechas };
}

/* ------------------------------------------------------------- */
/* Prompt: 1 diagnóstico + 1 examen + justificación basada en pts */
/* ------------------------------------------------------------- */

function construirMensajesIA(p) {
  const info = {
    nombre: p?.nombre || "",
    rut: p?.rut || "",
    edad: p?.edad || "",
    genero: p?.genero || "",
    dolor: p?.dolor || "",
    lado: p?.lado || "",
  };

  const { texto: puntosTexto } = describirPuntosClinicos(p?.detalles || null);

  const system = [
    "Eres un asistente clínico de traumatología e imagenología.",
    "Selecciona UN diagnóstico presuntivo y UN examen imagenológico inicial.",
    "La justificación clínica DEBE referenciar explícitamente los puntos dolorosos marcados.",
    "Evita sobreestudio: prioriza RX o ecografía si resuelven la pregunta clínica.",
    "Incluye lateralidad textual (izquierda/derecha) cuando aplique.",
    "Responde SIEMPRE SOLO JSON válido, sin texto fuera del JSON.",
  ].join(" ");

  const user = `
PACIENTE
${JSON.stringify(info, null, 2)}

HALLAZGOS DIRIGIDOS (PUNTOS DOLOROSOS)
${puntosTexto}

SALIDA (SOLO JSON VÁLIDO)
{
  "diagnostico_presuntivo": "3–8 palabras, específico, con lateralidad si aplica",
  "examen": "UN SOLO EXAMEN, EN MAYÚSCULAS",
  "justificacion_clinica": "60–120 palabras, referenciando los puntos y su correlato anatómico/patológico"
}
`.trim();

  return [
    { role: "system", content: system },
    { role: "user", content: user },
  ];
}

/* ------------------------------------------------------------- */
/* Llamada al modelo                                              */
/* ------------------------------------------------------------- */

async function llamarIA(mensajes) {
  const key = process.env.OPENAI_API_KEY || "";
  if (!key) throw new Error("OPENAI_API_KEY no configurada");

  const r = await fetch(OPENAI_API, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      authorization: `Bearer ${key}`,
    },
    body: JSON.stringify({
      model: MODEL,
      temperature: 0.35,
      response_format: { type: "json_object" },
      messages,
    }),
  });

  if (!r.ok) {
    const raw = await r.text().catch(() => "");
    throw new Error(`OpenAI ${r.status}: ${raw}`);
  }
  const j = await r.json();
  const txt = j?.choices?.[0]?.message?.content || "{}";
  return JSON.parse(txt);
}

/* ------------------------------------------------------------- */
/* Handler principal                                              */
/* ------------------------------------------------------------- */

export default function traumaIAHandler(memoria) {
  const ns = (s, id) => `${s}:${id}`;

  return async (req, res) => {
    try {
      const {
        idPago,
        paciente = {},
        detalles = null,
        rodillaMarcadores = null, // ← también aceptamos top-level
      } = req.body || {};
      if (!idPago) return res.status(400).json({ ok: false, error: "Falta idPago" });

      // Normaliza: preferimos detalles.rodillaMarcadores; si no, usamos top-level
      const detallesAll = {
        ...(detalles || {}),
        rodillaMarcadores: (detalles?.rodillaMarcadores ?? rodillaMarcadores) || null,
      };
      const p = { ...paciente, detalles: detallesAll };

      // IA con puntos
      let out;
      try {
        const mensajes = construirMensajesIA(p);
        const ia = await llamarIA(mensajes);

        const diagnostico = String(ia?.diagnostico_presuntivo || "").trim();
        const examenRaw  = String(ia?.examen || "").trim();
        const justificacion = String(ia?.justificacion_clinica || "").trim();

        const examen = normalizarExamen(examenRaw, p?.dolor, p?.lado);

        if (!diagnostico || !examen) {
          out = fallbackHeuristico(p);
        } else {
          out = { diagnostico, examen, justificacion };
        }
      } catch {
        out = fallbackHeuristico(p);
      }

      // Persistencia mínima (si usas memoria para PDF/reportes)
      const registro = {
        ...p,
        examenesIA: [out.examen],
        respuesta: `Diagnóstico presuntivo: ${out.diagnostico}\n\n${out.justificacion}`,
        pagoConfirmado: true,
      };
      try {
        memoria?.set?.(ns("ia", idPago), registro);
        memoria?.set?.(ns("meta", idPago), { moduloAutorizado: "ia" });
      } catch {}

      // Respuesta final
      return res.json({
        ok: true,
        diagnostico: out.diagnostico,
        examenes: [out.examen],
        justificacion: out.justificacion,
      });
    } catch (e) {
      console.error("ia-trauma error:", e);
      return res.status(500).json({ ok: false, error: e.message });
    }
  };
}

File: ia/trauma_ia.py
# ia/trauma_ia.py
# TRAUMA IA — nota breve estricta: 1 diagnóstico + 1 examen
# Claude primario → OpenAI fallback → heurístico
# Stateless — sin memoria server-side
# CONTRATO: { ok, diagnostico, examenes[1], justificacion, informeIA }

import re
import logging
from typing import Any

import httpx

from fallbacks.trauma import fallback_trauma

logger = logging.getLogger("trauma_ia")

# ============================================================
# SYSTEM PROMPT
# ============================================================
SYSTEM_PROMPT = """
Eres un asistente clínico de TRAUMATOLOGÍA para pre-orientación.
Objetivo: redactar una NOTA BREVE centrada en EXÁMENES a solicitar.

Reglas (ESTRICTAS):
- Español claro. Extensión total: 140–170 palabras.
- NO es diagnóstico definitivo ni tratamiento. No prescribas fármacos.
- Evita alarmismo. Usa condicionales ("podría sugerir", "compatible con").
- Prioriza IMAGENOLOGÍA. Si corresponde, sugiere ECOGRAFÍA en lesiones de partes blandas (p. ej., hombro/codo/mano en pacientes jóvenes).
- Si hay lateralidad (Derecha/Izquierda), inclúyela explícitamente en el examen.
- Integra PUNTOS DOLOROSOS si existen; la explicación debe referirse a ellos cuando estén presentes.
- **EXACTAMENTE 1** diagnóstico presuntivo.
- **EXACTAMENTE 1** examen sugerido.
- No repitas identificadores del paciente.

Formato EXACTO (mantén títulos y viñetas tal cual):
Diagnóstico presuntivo:
• (una sola entidad clínica específica a la zona)

Explicación breve:
• (≈60–100 palabras, 1–3 frases que justifiquen el enfoque y el porqué del examen; referencia a los puntos dolorosos si existen)

Exámenes sugeridos:
• (UN SOLO EXAMEN — incluir lateralidad si aplica)

Indicaciones:
• Presentarse con la orden; ayuno solo si el examen lo solicita.
• Acudir a evaluación presencial con el/la especialista sugerido/a.

Devuelve SOLO el texto en este formato (sin comentarios adicionales).
""".strip()


# ============================================================
# MARCADORES — helpers retro-compat
# ============================================================
def _slug(s: str) -> str:
    return str(s).strip().lower()


def _san_vista(obj: dict) -> dict:
    norm: dict[str, list[str]] = {}
    for vista in ["frente", "lateral", "posterior"]:
        arr = obj.get(vista, [])
        norm[vista] = [str(x).strip() for x in (arr if isinstance(arr, list) else []) if str(x).strip()]
    for k, v in obj.items():
        if k not in norm and isinstance(v, list):
            norm[k] = [str(x).strip() for x in v if str(x).strip()]
    return norm


def _leer_marcadores_desde_body(body: dict) -> dict:
    out: dict[str, dict] = {}
    marcadores = body.get("marcadores")
    if isinstance(marcadores, dict):
        for region, obj in marcadores.items():
            if isinstance(obj, dict):
                out[_slug(region)] = _san_vista(obj)
    for k, v in body.items():
        m = re.match(r'^([a-zA-ZñÑ]+)Marcadores$', k)
        if m and isinstance(v, dict):
            out[_slug(m.group(1))] = _san_vista(v)
    return out


def _filtrar_regiones_relevantes(marcadores: dict, dolor: str = "") -> dict:
    regiones = list(marcadores.keys())
    if not regiones:
        return {}
    d = str(dolor or "").lower()
    hits = [r for r in regiones if r in d]
    return {r: marcadores[r] for r in hits} if hits else marcadores


def _uc(s: str) -> str:
    return s[0].upper() + s[1:] if s else s


def _marcadores_a_texto(m_reg: dict) -> str:
    bloques = []
    for region, vistas in m_reg.items():
        sub = []
        for vista, arr in vistas.items():
            if isinstance(arr, list) and arr:
                sub.append(f"{_uc(vista)}:\n• " + "\n• ".join(arr))
        if sub:
            bloques.append(f"{_uc(region)} — Puntos marcados\n" + "\n\n".join(sub))
    return "\n\n".join(bloques) if bloques else "Sin puntos dolorosos marcados."


def _flat(obj: dict) -> list[str]:
    out = []
    for v in obj.values():
        if isinstance(v, list):
            out.extend(v)
    return [str(s).lower() for s in out]


def _tips_rodilla(obj: dict) -> list[str]:
    t = _flat(obj)
    has = lambda rx: any(re.search(rx, x) for x in t)
    arr = []
    if has(r'\binterl[ií]nea?\s+medial\b'):    arr.append("Interlínea medial → sospecha menisco medial.")
    if has(r'\binterl[ií]nea?\s+lateral\b'):   arr.append("Interlínea lateral → sospecha menisco lateral.")
    if has(r'\b(r[óo]tula|patelar|patelofemoral|ap[eé]x)\b'): arr.append("Dolor patelofemoral → síndrome PF/condropatía.")
    if has(r'\btuberosidad\s+tibial\b'):        arr.append("Tuberosidad tibial → Osgood–Schlatter / tendón rotuliano.")
    if has(r'\b(pes\s+anserin[oó]|pata\s+de\s+ganso)\b'): arr.append("Pes anserino → tendinopatía/bursitis anserina.")
    if has(r'\b(gerdy|banda\s+ilio?tibial|tracto\s+ilio?tibial)\b'): arr.append("Banda ITB/Gerdy → síndrome banda ITB.")
    if has(r'\bpopl[ií]tea?\b'):                arr.append("Fosa poplítea → evaluar quiste de Baker.")
    return arr


def _tips_hombro(obj: dict) -> list[str]:
    t = _flat(obj)
    has = lambda rx: any(re.search(rx, x) for x in t)
    arr = []
    if has(r'\b(subacromial|acromion|bursa\s*subacromial)\b'):           arr.append("Dolor subacromial → síndrome subacromial / supraespinoso.")
    if has(r'\b(tub[eé]rculo\s*mayor|footprint|troquiter)\b'):           arr.append("Tubérculo mayor → tendinopatía del manguito (supra/infra).")
    if has(r'\b(surco\s*bicipital|bicipital|porci[oó]n\s*larga\s*del\s*b[ií]ceps)\b'): arr.append("Surco bicipital → tendinopatía de la porción larga del bíceps.")
    if has(r'\b(acromioclavicular|acromio\-?clavicular)\b'):              arr.append("Dolor AC → artropatía acromioclavicular.")
    if has(r'\b(posterosuperior|labrum\s*superior|slap)\b'):              arr.append("Dolor posterosuperior → considerar lesión labral (SLAP).")
    return arr


def _tips_desde_marcadores(m_reg: dict) -> list[str]:
    tips = []
    for region, obj in m_reg.items():
        if region == "rodilla": tips.extend(_tips_rodilla(obj))
        if region == "hombro":  tips.extend(_tips_hombro(obj))
    return tips


# ============================================================
# NORMALIZAR EXAMEN
# ============================================================
def _normalizar_examen(examen: str, dolor: str = "", lado: str = "") -> str:
    x = str(examen or "").strip()
    if not x:
        return ""
    x = x.upper()
    if not x.endswith("."):
        x += "."
    lado_u = str(lado or "").upper()
    lat    = f" {lado_u}" if lado_u else ""
    if (
        re.search(r'\b(CADERA|RODILLA|HOMBRO|TOBILLO|PIERNA|BRAZO|CODO|MUÑECA|MANO|PIE)\b', x)
        and lat
        and not re.search(r'\b(IZQUIERDA|DERECHA)\b', x)
    ):
        x = x.rstrip(".") + lat + "."
    if re.search(r'ECOGRAF[ÍI]A.*PARTES\s+BLANDAS', x) and "DE" not in x:
        zona = str(dolor or "").upper()
        if zona and "COLUMNA" not in zona:
            x = f"ECOGRAFÍA DE PARTES BLANDAS DE {zona}{lat}.".upper()
    return x


# ============================================================
# FALLBACK HEURÍSTICO
# ============================================================
def _fallback_heuristico(p: dict) -> dict:
    return fallback_trauma(p)


# ============================================================
# CONSTRUIR MENSAJE USUARIO
# ============================================================
def _construir_mensaje_usuario(p: dict) -> str:
    dolor  = str(p.get("dolor") or "")
    lado   = str(p.get("lado") or "")
    edad   = p.get("edad")
    genero = str(p.get("genero") or "")
    marc   = (p.get("detalles") or {}).get("marcadores") or {}

    puntos_txt = _marcadores_a_texto(marc)
    tips_arr   = _tips_desde_marcadores(marc)
    tips_txt   = ("\n\nTips clínicos:\n• " + "\n• ".join(tips_arr)) if tips_arr else ""

    return (
        f"Edad: {edad or '—'}\n"
        + (f"Género: {genero}\n" if genero else "")
        + (f"Región de dolor: {dolor}{f' ({lado})' if lado else ''}\n" if dolor else "")
        + f"Puntos dolorosos marcados:\n{puntos_txt}{tips_txt}\n\n"
        + "Redacta EXACTAMENTE con el formato solicitado y el carácter ESTRICTO de 1 diagnóstico y 1 examen."
    )


# ============================================================
# PARSE SECCIONES
# ============================================================
def _parse_secciones(text: str) -> dict:
    out = {"diagnostico": "", "explicacion": "", "examen": ""}
    if not text:
        return out

    def _bullets(block: str) -> list[str]:
        result = []
        for linea in block.splitlines():
            linea = linea.strip()
            if not linea:
                continue
            m = re.match(r'^[•\-\*]\s*(.+)$', linea)
            result.append((m.group(1) if m else linea).strip())
        return [s for s in result if s]

    sec_dx = re.search(r'Diagn[oó]stico presuntivo:\s*([\s\S]*?)(?:\n\s*Explicaci[oó]n breve:|$)', text, re.I)
    if sec_dx:
        out["diagnostico"] = (_bullets(sec_dx.group(1)) or [""])[0]

    sec_exp = re.search(r'Explicaci[oó]n breve:\s*([\s\S]*?)(?:\n\s*Ex[aá]menes sugeridos:|$)', text, re.I)
    if sec_exp:
        out["explicacion"] = " ".join(_bullets(sec_exp.group(1))).strip()

    sec_ex = re.search(r'Ex[aá]men(?:es)? sugeridos?:\s*([\s\S]*?)(?:\n\s*Indicaciones:|$)', text, re.I)
    if sec_ex:
        bullets = [re.sub(r'\s*\.\s*$', '.', b) for b in _bullets(sec_ex.group(1))]
        out["examen"] = bullets[0] if bullets else ""

    return out


# ============================================================
# LLAMADAS IA
# ============================================================
async def _llamar_claude(user_msg: str, api_key: str, model: str) -> str:
    async with httpx.AsyncClient(timeout=30.0) as client:
        r = await client.post(
            "https://api.anthropic.com/v1/messages",
            headers={
                "x-api-key":         api_key,
                "anthropic-version": "2023-06-01",
                "content-type":      "application/json",
            },
            json={
                "model":      model,
                "max_tokens": 600,
                "system":     SYSTEM_PROMPT,
                "messages":   [{"role": "user", "content": user_msg}],
            },
        )
        r.raise_for_status()
        return (r.json().get("content") or [{}])[0].get("text", "").strip()


async def _llamar_openai(user_msg: str, api_key: str, model: str) -> str:
    async with httpx.AsyncClient(timeout=30.0) as client:
        r = await client.post(
            "https://api.openai.com/v1/chat/completions",
            headers={
                "authorization": f"Bearer {api_key}",
                "content-type":  "application/json",
            },
            json={
                "model":       model,
                "temperature": 0.35,
                "max_tokens":  520,
                "messages": [
                    {"role": "system", "content": SYSTEM_PROMPT},
                    {"role": "user",   "content": user_msg},
                ],
            },
        )
        r.raise_for_status()
        return (r.json().get("choices") or [{}])[0].get("message", {}).get("content", "").strip()


# ============================================================
# HANDLER PRINCIPAL
# ============================================================
async def trauma_ia(payload: dict, config: dict) -> dict:
    """
    payload: {
        idPago, paciente?, detalles?, traumaJSON?
    }
    config: {
        anthropic_api_key?, openai_api_key?,
        anthropic_model?,   openai_model?
    }
    Retorna: { ok, diagnostico, examenes[1], justificacion, informeIA }
    """
    id_pago      = payload.get("idPago") or ""
    paciente     = payload.get("paciente") or {}
    detalles     = payload.get("detalles") or {}
    trauma_json  = payload.get("traumaJSON")

    if not id_pago:
        return {"ok": False, "error": "Falta idPago"}

    marcadores_all = _leer_marcadores_desde_body(payload)

    if isinstance(trauma_json, dict):
        pac_tj = trauma_json.get("paciente") or {}
        paciente = {**paciente, **pac_tj}
        marc_tj  = trauma_json.get("marcadores") or {}
        if marc_tj:
            marcadores_all = marc_tj
        marc_relev = _filtrar_regiones_relevantes(
            marcadores_all, pac_tj.get("dolor") or paciente.get("dolor") or ""
        )
        detalles = {
            **detalles,
            "marcadores": marc_relev,
            "resonancia": trauma_json.get("resonancia"),
        }
    else:
        marc_relev = _filtrar_regiones_relevantes(
            marcadores_all, paciente.get("dolor") or ""
        )
        detalles = {**detalles, "marcadores": marc_relev}

    p = {**paciente, "detalles": detalles}

    anthropic_key   = config.get("anthropic_api_key") or ""
    openai_key      = config.get("openai_api_key") or ""
    anthropic_model = config.get("anthropic_model") or "claude-sonnet-4-6"
    openai_model    = config.get("openai_model") or "gpt-4o-mini"

    texto_ia  = ""
    proveedor = "fallback"
    out: dict | None = None

    user_msg = _construir_mensaje_usuario(p)

    # — Claude primario —
    if anthropic_key:
        try:
            texto_ia  = await _llamar_claude(user_msg, anthropic_key, anthropic_model)
            proveedor = "claude"
        except Exception as e:
            logger.warning("Claude falló trauma_ia: %s", e)

    # — OpenAI fallback —
    if not texto_ia and openai_key:
        try:
            texto_ia  = await _llamar_openai(user_msg, openai_key, openai_model)
            proveedor = "openai"
        except Exception as e:
            logger.warning("OpenAI falló trauma_ia: %s", e)

    if texto_ia:
        parsed = _parse_secciones(texto_ia)
        dx_ok  = str(parsed["diagnostico"]).strip()
        ex_ok  = _normalizar_examen(parsed["examen"], p.get("dolor"), p.get("lado"))
        just   = parsed["explicacion"] or "Justificación clínica basada en región y puntos dolorosos."
        if dx_ok and ex_ok:
            out = {"diagnostico": dx_ok, "examen": ex_ok, "justificacion": just}

    if not out:
        out = _fallback_heuristico(p)
        proveedor = "fallback"

    return {
        "ok":           True,
        "diagnostico":  out["diagnostico"],
        "examenes":     [out["examen"]],
        "justificacion": out["justificacion"],
        "informeIA":    out["justificacion"],
        "_debug": {
            "texto_bruto": texto_ia,
            "proveedor":   proveedor,
        },
  }
  
